beacon_node/execution_layer/src/engines.rs

```diff
@@ -179,164 +167,112 @@ impl Engine {
+                    );
+
                     // Send the node our latest forkchoice_state.
                     self.send_latest_forkchoice_state().await;
-
-                    *state_lock = EngineState::Synced
+                } else {
+                    debug!(
+                        self.log,
+                        "Execution engine online";
+                    );
                 }
-                Err(EngineApiError::IsSyncing) => {
-                    if logging.is_enabled() {
-                        warn!(
-                            self.log,
-                            "Execution engine syncing";
-                        )
-                    }
-
-                    // Send the node our latest forkchoice_state, it may assist with syncing.
-                    self.send_latest_forkchoice_state().await;
 
-                    *state_lock = EngineState::Syncing
-                }
-                Err(EngineApiError::Auth(err)) => {
-                    if logging.is_enabled() {
-                        warn!(
-                            self.log,
-                            "Failed jwt authorization";
-                            "error" => ?err,
-                        );
-                    }
-
-                    *state_lock = EngineState::AuthFailed
-                }
-                Err(e) => {
-                    if logging.is_enabled() {
-                        warn!(
-                            self.log,
-                            "Execution engine offline";
-                            "error" => ?e,
-                        )
-                    }
-                }
+                *state = EngineState::Synced;
+                *state
             }
-        }
-
-        if *state_lock != EngineState::Synced && logging.is_enabled() {
-            crit!(
-                self.log,
-                "No synced execution engines";
-            )
-        }
-    }
+            Err(EngineApiError::IsSyncing) => {
+                let mut state = self.state.write().await;
+                *state = EngineState::Syncing;
+                *state
+            }
+            Err(EngineApiError::Auth(err)) => {
+                error!(
+                    self.log,
+                    "Failed jwt authorization";
+                    "error" => ?err,
+                );
 
-    /// Run `func` on the node.
-    ///
-    /// This function might try to run `func` twice. If the node returns an error it will try to
-    /// upcheck it and then run the function again.
-    pub async fn first_success<'a, F, G, H>(&'a self, func: F) -> Result<H, EngineError>
-    where
-        F: Fn(&'a Engine) -> G + Copy,
-        G: Future<Output = Result<H, EngineApiError>>,
-    {
-        match self.first_success_without_retry(func).await {
-            Ok(result) => Ok(result),
+                let mut state = self.state.write().await;
+                *state = EngineState::AuthFailed;
+                *state
```

Comment:
I'm more asking the for my own knowledge - is this correct?

1. As long as you hold `state` inside this branch, you hold the write lock
2. Since the call to the match function is `let state = match ...` and not `let mut state = match ...`, the write lock is dropped upon leaving this branch?
3. If the call *was* `let mut state = match ...` the write lock would **not** be dropped upon exiting this branch until the outer `state` was dropped?

Reply:
Oh yeah, good question! I'll try answer these questions in a round-about way:

### Example 1: Deref

```rust
let state: EngineState = match thing {
  Ok(()) => {
    let mut state = self.state.write().await;
    *state
  },
  Err(e) => unimplemented!()
}
```

### Example 2: No deref

```rust
let state: RwLockWriteGuard<EngineState> = match thing {
  Ok(()) => {
    let mut state = self.state.write().await;
    state
  },
  Err(e) => unimplemented!()
}
```

The key difference between these two examples is the [dereferencing](https://docs.rs/tokio/0.2.13/tokio/sync/struct.RwLockWriteGuard.html#impl-Deref) the `RwLockWriteGuard<EngineState>` to `EngineState`. I.e. `state` vs `*state`.

In example 1, we dereference "out of" the `RwLockWriteGuard` and just return a `Copy` of the EngineState, causing the write-guard to be dropped.

In example 2 we return the write-guard without ever dropping it.

In eb285d51b I've added an explicit type based on your comments. I think it improves readability. It also makes it more difficult to accidentally start returning a reference to the lock later on (not that it *really* matters with the current code).

Reply:
ah thanks for the explanation!